integration/teardown/teardown.go

```diff
@@ -3,39 +3,149 @@
 package teardown
 
 import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/backoff"
 	"github.com/giantswarm/e2e-harness/pkg/framework"
+	"github.com/giantswarm/kvm-operator/integration/utils"
 	"github.com/giantswarm/microerror"
-
-	"github.com/giantswarm/kvm-operator/integration/env"
+	"github.com/giantswarm/micrologger"
 )
 
 // Teardown e2e testing environment.
 func Teardown(g *framework.Guest, h *framework.Host) error {
 	var err error
+	var l micrologger.Logger
+	{
+		c := micrologger.Config{}
 
+		l, err = micrologger.New(c)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+	clusterID := getClusterID(h.TargetNamespace())
+
+	// get flannel info so we can delete it from rangepool
+	var flannelNetwork string
 	{
-		// only do full teardown when not on CI
-		if env.CircleCI() == "true" {
-			return nil
+		flannelConfig, err := h.G8sClient().CoreV1alpha1().FlannelConfigs(v1.NamespaceDefault).Get(clusterID, v1.GetOptions{})
+		if err != nil {
+			return microerror.Mask(err)
 		}
+		flannelNetwork = flannelConfig.Spec.Flannel.Spec.Network
 	}
 
 	{
-		err = framework.HelmCmd("delete kvm-operator --purge")
+		err = framework.HelmCmd(fmt.Sprintf("delete %s-cert-config-e2e --purge", h.TargetNamespace()))
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		err = framework.HelmCmd(fmt.Sprintf("delete %s-flannel-config-e2e --purge", h.TargetNamespace()))
 		if err != nil {
 			return microerror.Mask(err)
 		}
-		err = framework.HelmCmd("delete cert-operator --purge")
+		err = framework.HelmCmd(fmt.Sprintf("delete %s-kvm-config-e2e --purge", h.TargetNamespace()))
 		if err != nil {
 			return microerror.Mask(err)
 		}
 	}
 
+	// wait until crds are deleted by operators
+	o := func() error {
+		kvmDeleted, flannelDeleted, certDeleted := false, false, false
+
+		certList, err := h.G8sClient().CoreV1alpha1().CertConfigs(v1.NamespaceDefault).List(v1.ListOptions{
+			LabelSelector: crdLabelSelector(clusterID),
+		})
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		if len(certList.Items) == 0 {
+			// resource doesnt exist, we are good to continue
+			certDeleted = true
+			l.Log("level", "info", "message", "cert crd was deleted")
+		} else {
+			l.Log("level", "info", "message", "cert crd has not been deleted")
```

Comment:
You could just terminate early here with `return resourceNotDeleted` IMO. Then you won't need the variables and all three have to be deleted anyway, right?

Uncertain if this is necessary to change, just something I noticed.

Reply:
Good idea, will fix it.